packages/token-bridge-sdk/src/hooks/useBalance.ts

```diff
@@ -1,47 +1,134 @@
+import { MultiCaller } from '@arbitrum/sdk'
 
 import { useChainId } from './useChainId'
 
+type Erc20Balances = {
+  [address: string]: BigNumber | undefined
+}
+
+const merge: Middleware = useSWRNext => {
+  return (key, fetcher, config) => {
+    const { cache } = useSWRConfig()
+
+    const extendedFetcher = async () => {
+      const newBalances = await fetcher!()
+      const oldData = cache.get(unstable_serialize(key))
+      return {
+        ...oldData,
+        ...newBalances
+      }
+    }
+
+    return useSWRNext(key, extendedFetcher, config)
+  }
+}
+
 const useBalance = ({
   provider,
-  walletAddress
+  walletAddress,
+  erc20Addresses
 }: {
   provider: providers.Provider
   walletAddress: string | undefined
+  erc20Addresses?: string[]
 }) => {
   const chainId = useChainId({ provider })
   const walletAddressLowercased = useMemo(
     () => walletAddress?.toLowerCase(),
     [walletAddress]
   )
 
-  const queryKey = useMemo(() => {
-    if (
-      typeof chainId === 'undefined' ||
-      typeof walletAddressLowercased === 'undefined'
-    ) {
-      // Don't fetch
-      return null
+  const queryKey = useCallback(
+    (type: 'eth' | 'erc20') => {
+      if (
+        typeof chainId === 'undefined' ||
+        typeof walletAddressLowercased === 'undefined'
+      ) {
+        // Don't fetch
+        return null
+      }
+
+      return [chainId, walletAddressLowercased, 'balance', type] as const
+    },
+    [chainId, walletAddressLowercased]
+  )
+  const fetchErc20 = useCallback(async () => {
+    if (!queryKey('erc20') || !erc20Addresses) {
+      return {}
     }
 
-    return [chainId, walletAddressLowercased, 'ethBalance']
-  }, [chainId, walletAddressLowercased])
+    const multiCaller = await MultiCaller.fromProvider(provider as Provider)
+    const addressesBalances = await multiCaller.getTokenData(erc20Addresses, {
+      balanceOf: { account: walletAddressLowercased! }
+    })
+
+    return erc20Addresses.reduce((acc, address, index) => {
+      const balance = addressesBalances[index]
+      if (balance) {
+        acc[address.toLowerCase()] = balance.balance
+      }
 
-  const { data: dataEth = null, mutate: mutateEth } = useSWR(
-    queryKey,
-    (_, _walletAddress: string) => provider.getBalance(_walletAddress),
+      return acc
+    }, {} as Erc20Balances)
+  }, [queryKey, erc20Addresses, provider, walletAddressLowercased])
+
+  const { data: dataEth = null, mutate: updateEthBalance } = useSWR(
+    queryKey('eth'),
+    (_, _walletAddress) => provider.getBalance(_walletAddress),
     {
       refreshInterval: 15_000,
       shouldRetryOnError: true,
       errorRetryCount: 2,
       errorRetryInterval: 3_000
     }
   )
+  const { data: dataErc20 = null, mutate: mutateErc20 } = useSWR(
+    queryKey('erc20'),
+    fetchErc20,
+    {
+      shouldRetryOnError: true,
+      errorRetryCount: 2,
+      errorRetryInterval: 3_000,
+      use: [merge]
+    }
+  )
+
+  const updateErc20 = useCallback(
```

Comment:
We discussed this privately -- ideally the API should just allow you to provide the list of addresses you want the balance updated for, so the client doesn't have to provide the new balance values